### ISSUE TITLE
The implementation of WebView SetVirtualHostNameToFolderMapping on Android is not aligned with WinUI3

### ISSUE THREAD

**Issue by ForkBug** [unoplatform/uno#18930]
### Current behavior

The virtual hostname is mapped for navigation urls, but not for web resources(css, js, etc.)
For `web.CoreWebView2.SetVirtualHostNameToFolderMapping(
                    "anyhost",
                    "Htmls",
                    CoreWebView2HostResourceAccessKind.Allow);`
A html file containing `<img src="http://anyhost/img.png">` will fail to load the image on Android, but it works as expected on Windows 11.

### Expected behavior

The hostname mapped by SetVirtualHostNameToFolderMapping gets resolved when loading CSS and images.

### How to reproduce it (as minimally and precisely as possible)

The issue is in https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/Controls/WebView/Native/Android/InternalWebClient.Android.cs

InternalClient needs to implement [shouldInterceptRequest](https://developer.android.com/reference/android/webkit/WebViewClient#shouldInterceptRequest(android.webkit.WebView,%20android.webkit.WebResourceRequest)) and then lookup _coreWebView.HostToFolderMap

### Workaround

No workaround

### Works on UWP/WinUI

None

### Environment

Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia

### NuGet package version(s)

    "Uno.Sdk": "5.5.43"

### Affected platforms

Android

### IDE

Visual Studio 2022

### IDE version

2022 17.12

### Relevant plugins

Not relative

### Anything else we need to know?

Does the issue exist on Ios and Macos? I haven't tested it yet.

**Comment by Youssef1313** [comment]
Yes, I noticed that when I previously worked on WebView support for WPF. To me, it seemed like a limitation of the native Android API. The only way around this is that we alter the HTML ourselves, but not trivial to "correctly" support all the scenarios as in WinUI. Note: for WPF. I added a new interface, `ISupportsVirtualHostMapping` that signifies a "proper" support for the feature.

**Comment by ForkBug** [comment]
> Yes, I noticed that when I previously worked on WebView support for WPF. To me, it seemed like a limitation of the native Android API. The only way around this is that we alter the HTML ourselves, but not trivial to "correctly" support all the scenarios as in WinUI. Note: for WPF. I added a new interface, `ISupportsVirtualHostMapping` that signifies a "proper" support for the feature. The other platforms which don't use the interface are falling back to a "best effort" implementation which is not very correct.

The implementation is not complex but I'm not familiar with Uno dev process and the current architecture of Uno Webview. Some demonstration code(toy code):

`
        public override NSAndroid.Webkit.WebResourceResponse? ShouldInterceptRequest(NSAndroid.Webkit.WebView? view,
            NSAndroid.Webkit.IWebResourceRequest? request)
        {

            if (request?.Url?.Scheme?.Equals("file", StringComparison.OrdinalIgnoreCase) == true)
            {
                var filePath = NSAndroid.Net.Uri.Parse(request.Url.ToString())?.Path ?? "";
                filePath = Path.Combine(Mapped, filePath);
                if (File.Exists(filePath))
                {
                    string mimeType = GetMimeType(filePath);
                    using var stream = File.OpenRead(filePath);
                    return new NSAndroid.Webkit.WebResourceResponse(mimeType, null, stream);
                }
            }
            return base.ShouldInterceptRequest(view, request);
        }`

**Comment by Youssef1313** [comment]
I didn't came across `ShouldInterceptRequest`, sounds promising. cc @MartinZikmund

**Comment by jeromelaban** [comment]
This like should be implemented around here:

https://github.com/unoplatform/uno/blob/58e477bd978a209f7e630fde1fae43acf1d4330d/src/Uno.UI/UI/Xaml/Controls/WebView/Native/Android/NativeWebView.Android.cs#L6-L6

and here:

https://github.com/unoplatform/uno/blob/58e477bd978a209f7e630fde1fae43acf1d4330d/src/Uno.UI/UI/Xaml/Controls/WebView/Native/Android/NativeWebViewWrapper.Android.cs#L17-L17

**Comment by ForkBug** [comment]
> This like should be implemented around here:
> 
> https://github.com/unoplatform/uno/blob/58e477bd978a209f7e630fde1fae43acf1d4330d/src/Uno.UI/UI/Xaml/Controls/WebView/Native/Android/NativeWebView.Android.cs#L6-L6
> 
> and here:
> 
> https://github.com/unoplatform/uno/blob/58e477bd978a209f7e630fde1fae43acf1d4330d/src/Uno.UI/UI/Xaml/Controls/WebView/Native/Android/NativeWebViewWrapper.Android.cs#L17-L17

There are 2 scenes the hostname is used:

1. Navigation. The web view loads an HTML/URL, or users click a link that causes the web view to load another HTML/URL. The HTML file is in the top frame of the web view.
2. Web resources(css/js/imags...) loading. An HTML file is loaded into web view and refers to some css/img/js... files in another mapped hostname. The HTML may be in "anyhost1" but the image files may be in "anyhost2". The HTML contains some node like `<img src="http://anyhost2/img.png">`

WinUI3 support 1 and 2. 
Uno Android only supports 1.

The root cause is in [[uno/src/Uno.UI/UI/Xaml/Controls/WebView/Native/Android/NativeWebViewWrapper.Android.cs](https://github.com/unoplatform/uno/blob/58e477bd978a209f7e630fde1fae43acf1d4330d/src/Uno.UI/UI/Xaml/Controls/WebView/Native/Android/NativeWebViewWrapper.Android.cs#L17-L17)](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/Controls/WebView/Native/Android/InternalWebClient.Android.cs)

To support scene 1, InternalClient overrides ShouldOverrideUrlLoading and then calls ProcessNavigation in [uno/src/Uno.UI/UI/Xaml/Controls/WebView/Native/Android/NativeWebViewWrapper.Android.cs](https://github.com/unoplatform/uno/blob/58e477bd978a209f7e630fde1fae43acf1d4330d/src/Uno.UI/UI/Xaml/Controls/WebView/Native/Android/NativeWebViewWrapper.Android.cs#L17-L17)

To support scene 2, InternalClient has to override [shouldInterceptRequest](https://developer.android.com/reference/android/webkit/WebViewClient#shouldInterceptRequest(android.webkit.WebView,%20android.webkit.WebResourceRequest))
, and looks up _coreWebView.HostToFolderMap and returns the stream object in shouldInterceptRequest.

**Comment by ForkBug** [comment]
Due to the same issue, the HTML tag <base  href="/" /> in the main HTML doesn't work either.

**Comment by ForkBug** [comment]
For IOS, setURLSchemeHandler could be used for interpreting web resource loading.
But that would bring another problem: Web resources have to be in the form of <img src="Unooo://anyhost2/img.png">

**Comment by ForkBug** [comment]
Workaround:
Create a web server by HttpListener or embedio. 

A little slow but it works, with more flexibility.
